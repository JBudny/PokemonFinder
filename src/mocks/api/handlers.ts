import { rest } from 'msw'

import { dummyPokemon } from '@mocks/dummyPokemon'

export const handlers = [
	rest.get('*/pokemon/:id', (req, res, ctx) => {
		return res(ctx.status(200), ctx.json(dummyPokemon))
	})
]
