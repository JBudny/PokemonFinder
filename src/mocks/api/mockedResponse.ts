import { dummyPokemon } from '@mocks/dummyPokemon'

export const mockedResponse = [dummyPokemon]
