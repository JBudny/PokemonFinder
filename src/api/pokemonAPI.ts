import Config from 'react-native-config'

import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'

import { PokemonItem } from './types/pokemonItem.types'
import { pokemonQueryFn } from './utils/pokemonQueryFn'

const { POKEMON_API_BASE_URL } = Config

export const api = createApi({
	baseQuery: fetchBaseQuery({ baseUrl: POKEMON_API_BASE_URL }),
	keepUnusedDataFor: process.env.NODE_ENV === 'test' ? 0 : 60,
	endpoints: (build) => ({
		pokemon: build.query<PokemonItem, string>({
			queryFn: pokemonQueryFn
		})
	})
})

export const { usePokemonQuery } = api
