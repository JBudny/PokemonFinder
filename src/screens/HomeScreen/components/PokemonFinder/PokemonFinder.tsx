import React, { FC, useState } from 'react'

import { skipToken } from '@reduxjs/toolkit/dist/query'

import { usePokemonQuery } from '@api/pokemonAPI'
import { Finder } from '@components/Finder/Finder'
import { LoadingIndicator } from '@components/LoadingIndicator/LoadingIndicator'
import { NoResults } from '@components/NoResults/NoResults'
import { useDebounce } from '@hooks/useDebounce/useDebounce'

import { PokemonCard } from './PokemonCard/PokemonCard'

/**
 * Instance of the Finder component.
 */
export const PokemonFinder: FC = () => {
	const [searchTerm, setSearchTerm] = useState<string | null>(null)
	const debouncedSearchTerm = useDebounce(searchTerm, 500)
	const { data, isError, isSuccess, isFetching } = usePokemonQuery(
		debouncedSearchTerm || skipToken
	)

	const handleChangeText = (text: string) => setSearchTerm(text)

	return (
		<Finder
			data={data}
			isSuccess={isSuccess}
			isFetching={isFetching}
			isError={isError}
			textInputProps={{
				onChangeText: handleChangeText,
				placeholder: 'Eg. bulbasaur'
			}}
			ItemComponent={PokemonCard}
			LoadingIndicator={LoadingIndicator}
			NoResults={NoResults}
		/>
	)
}
