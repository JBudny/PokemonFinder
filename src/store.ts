import type { PreloadedState } from '@reduxjs/toolkit'
import { combineReducers, configureStore } from '@reduxjs/toolkit'

import { api } from '@api/pokemonAPI'

const rootReducer = combineReducers({
	[api.reducerPath]: api.reducer
})

export const setupStore = (preloadedState?: PreloadedState<RootState>) =>
	configureStore({
		reducer: rootReducer,
		middleware: (getDefaultMiddleware) =>
			// adding the api middleware enables caching, invalidation, polling and other features of `rtk-query`
			getDefaultMiddleware().concat(api.middleware),
		preloadedState
	})

export type RootState = ReturnType<typeof rootReducer>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = AppStore['dispatch']
