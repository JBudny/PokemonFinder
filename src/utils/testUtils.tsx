import React from 'react'

import type { PreloadedState } from '@reduxjs/toolkit'
import type { RenderOptions } from '@testing-library/react-native'
import { render } from '@testing-library/react-native'

import type { AppStore, RootState } from '@src/store'
import { setupStore } from '@src/store'

import { Providers } from './Providers'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store. For
// future dependencies, such as wanting to test with react-router, you can extend
// this interface to accept a path and route and use those in a <MemoryRouter />
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
	preloadedState?: PreloadedState<RootState>
	store?: AppStore
}

const renderWithProviders = (
	ui: React.ReactElement,
	{
		preloadedState = {},
		store = setupStore(preloadedState),
		...renderOptions
	}: ExtendedRenderOptions = {}
) => {
	const Wrapper: React.FC = ({ children }) => (
		<Providers store={store}>{children}</Providers>
	)

	return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

export { renderWithProviders }
