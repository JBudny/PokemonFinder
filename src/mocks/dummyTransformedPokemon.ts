import { transformPokemon } from '@src/api/utils/transformPokemon'

import { dummyPokemon } from './dummyPokemon'

export const dummyTransformedPokemon = transformPokemon(dummyPokemon).data
