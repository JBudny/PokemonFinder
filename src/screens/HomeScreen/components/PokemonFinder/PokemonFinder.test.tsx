import React from 'react'
import Config from 'react-native-config'

import { act, fireEvent } from '@testing-library/react-native'
import { rest } from 'msw'

import { setupStore } from '@src/store'
import { server } from '@mocks/api/server'
import { dummyPokemon } from '@mocks/dummyPokemon'
import { dummyTransformedPokemon } from '@mocks/dummyTransformedPokemon'
import { renderWithProviders } from '@utils/testUtils'

import { PokemonFinder } from './PokemonFinder'

const { POKEMON_API_BASE_URL } = Config

const placeholder = 'Eg. bulbasaur'
const { sprite, name, abilities } = dummyTransformedPokemon

beforeEach(() => {
	jest.useFakeTimers()
})

afterEach(() => {
	jest.useRealTimers()
})

const delay = 500

test('PokemonFinder should display proper results', async () => {
	const store = setupStore()
	const { findByText, getByPlaceholderText, findAllByTestId } =
		renderWithProviders(<PokemonFinder />, { store })
	const searchInput = getByPlaceholderText(placeholder)
	fireEvent.changeText(searchInput, name)
	act(() => {
		jest.advanceTimersByTime(delay)
	})
	const nameText = await findByText(name)
	const abilitiesText = await findByText(abilities)
	const statsElements = await findAllByTestId('pokemon-stat')
	expect(nameText).toBeTruthy()
	expect(abilitiesText).toBeTruthy()
	expect(statsElements).toHaveLength(6)
})

test('PokemonFinder should display image with proper source', async () => {
	const store = setupStore()
	const { findByTestId, getByPlaceholderText } = renderWithProviders(
		<PokemonFinder />,
		{ store }
	)
	const searchInput = getByPlaceholderText(placeholder)
	fireEvent.changeText(searchInput, name)
	act(() => {
		jest.advanceTimersByTime(delay)
	})
	const image = await findByTestId('official-artwork')
	expect(image.props.source.uri).toBe(sprite)
})

test('PokemonFinder should display no image animation if sprite is null', async () => {
	server.use(
		rest.get(`${POKEMON_API_BASE_URL}/pokemon/*`, (_req, res, ctx) => {
			return res(
				ctx.json({
					...dummyPokemon,
					sprites: { other: { 'official-artwork': { front_default: null } } }
				})
			)
		})
	)
	const store = setupStore()
	const { getByPlaceholderText, findByTestId } = renderWithProviders(
		<PokemonFinder />,
		{ store }
	)
	const searchInput = getByPlaceholderText(placeholder)
	fireEvent.changeText(searchInput, name)
	act(() => {
		jest.advanceTimersByTime(delay)
	})
	const noImageAnimation = await findByTestId('no-image-animation')
	expect(noImageAnimation).toBeDefined()
})

test('PokemonFinder result should match the snapshot', async () => {
	const store = setupStore()
	const result = renderWithProviders(<PokemonFinder />, { store })
	const searchInput = result.getByPlaceholderText(placeholder)
	fireEvent.changeText(searchInput, name)
	act(() => {
		jest.advanceTimersByTime(delay)
	})
	await result.findByText(name)
	expect(result).toMatchSnapshot()
})
