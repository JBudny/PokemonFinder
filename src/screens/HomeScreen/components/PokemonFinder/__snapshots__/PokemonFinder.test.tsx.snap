// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonFinder result should match the snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "rgb(255, 255, 255)",
        "borderRadius": 15,
        "elevation": 8,
        "overflow": "hidden",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "rgb(255, 255, 255)",
          "borderRadius": 15,
          "flexDirection": "row",
          "minHeight": 48,
          "minWidth": 48,
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginRight": 10,
          },
        ]
      }
      testID="icon"
    >
      <
        color="rgb(150, 150, 150)"
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [
                128269,
                "search",
              ],
              "f002",
              "M500.3 443.7l-119.7-119.7c27.22-40.41 40.65-90.9 33.46-144.7C401.8 87.79 326.8 13.32 235.2 1.723C99.01-15.51-15.51 99.01 1.724 235.2c11.6 91.64 86.08 166.7 177.6 178.9c53.8 7.189 104.3-6.236 144.7-33.46l119.7 119.7c15.62 15.62 40.95 15.62 56.57 0C515.9 484.7 515.9 459.3 500.3 443.7zM79.1 208c0-70.58 57.42-128 128-128s128 57.42 128 128c0 70.58-57.42 128-128 128S79.1 278.6 79.1 208z",
            ],
            "iconName": "magnifying-glass",
            "prefix": "fas",
          }
        }
        size={20}
      />
    </View>
    <TextInput
      onChangeText={[Function]}
      placeholder="Eg. bulbasaur"
      placeholderTextColor="rgb(150, 150, 150)"
      style={
        Array [
          Object {
            "color": "rgb(33, 26, 29)",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontSize": 20,
            "paddingBottom": 0,
            "paddingLeft": 0,
            "paddingRight": 0,
            "paddingTop": 0,
          },
        ]
      }
    />
  </View>
  <View
    collapsable={false}
    style={
      Object {
        "overflow": "hidden",
        "transform": Array [
          Object {
            "translateY": -400,
          },
        ],
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "height": 550,
            "justifyContent": "space-between",
            "paddingBottom": 20,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 20,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "rgb(212, 241, 255)",
              "borderRadius": 150,
              "height": 200,
              "justifyContent": "center",
              "width": 200,
            },
          ]
        }
      >
        <Image
          onLoadEnd={[Function]}
          onLoadStart={[Function]}
          source={
            Object {
              "uri": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/35.png",
            }
          }
          style={
            Array [
              Object {
                "alignSelf": "center",
                "bottom": -40,
                "height": 250,
                "position": "absolute",
                "width": 250,
              },
            ]
          }
          testID="official-artwork"
        />
      </View>
      <Text
        size="large"
        style={
          Array [
            Object {
              "color": "rgb(255, 255, 255)",
              "fontSize": 35,
              "fontWeight": "bold",
            },
            Object {
              "color": "rgb(33, 26, 29)",
              "textTransform": "capitalize",
            },
          ]
        }
      >
        clefairy
      </Text>
      <Text
        size="medium"
        style={
          Array [
            Object {
              "color": "rgb(33, 26, 29)",
              "fontSize": 20,
              "fontWeight": "normal",
            },
          ]
        }
      >
        cute-charm, magic-guard, friend-guard
      </Text>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "flexWrap": "wrap",
              "justifyContent": "space-between",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
                "height": 30,
                "justifyContent": "space-between",
                "width": "48%",
              },
            ]
          }
          testID="pokemon-stat"
        >
          <Text
            size="small"
            style={
              Array [
                Object {
                  "color": "rgb(33, 26, 29)",
                  "fontSize": 16,
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "600",
                  "textTransform": "capitalize",
                },
              ]
            }
          >
            hp
          </Text>
          <Text
            size="small"
            style={
              Array [
                Object {
                  "color": "rgb(33, 26, 29)",
                  "fontSize": 16,
                  "fontWeight": "normal",
                },
              ]
            }
          >
            70
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
                "height": 30,
                "justifyContent": "space-between",
                "width": "48%",
              },
            ]
          }
          testID="pokemon-stat"
        >
          <Text
            size="small"
            style={
              Array [
                Object {
                  "color": "rgb(33, 26, 29)",
                  "fontSize": 16,
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "600",
                  "textTransform": "capitalize",
                },
              ]
            }
          >
            attack
          </Text>
          <Text
            size="small"
            style={
              Array [
                Object {
                  "color": "rgb(33, 26, 29)",
                  "fontSize": 16,
                  "fontWeight": "normal",
                },
              ]
            }
          >
            45
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
                "height": 30,
                "justifyContent": "space-between",
                "width": "48%",
              },
            ]
          }
          testID="pokemon-stat"
        >
          <Text
            size="small"
            style={
              Array [
                Object {
                  "color": "rgb(33, 26, 29)",
                  "fontSize": 16,
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "600",
                  "textTransform": "capitalize",
                },
              ]
            }
          >
            defense
          </Text>
          <Text
            size="small"
            style={
              Array [
                Object {
                  "color": "rgb(33, 26, 29)",
                  "fontSize": 16,
                  "fontWeight": "normal",
                },
              ]
            }
          >
            48
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
                "height": 30,
                "justifyContent": "space-between",
                "width": "48%",
              },
            ]
          }
          testID="pokemon-stat"
        >
          <Text
            size="small"
            style={
              Array [
                Object {
                  "color": "rgb(33, 26, 29)",
                  "fontSize": 16,
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "600",
                  "textTransform": "capitalize",
                },
              ]
            }
          >
            special-attack
          </Text>
          <Text
            size="small"
            style={
              Array [
                Object {
                  "color": "rgb(33, 26, 29)",
                  "fontSize": 16,
                  "fontWeight": "normal",
                },
              ]
            }
          >
            60
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
                "height": 30,
                "justifyContent": "space-between",
                "width": "48%",
              },
            ]
          }
          testID="pokemon-stat"
        >
          <Text
            size="small"
            style={
              Array [
                Object {
                  "color": "rgb(33, 26, 29)",
                  "fontSize": 16,
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "600",
                  "textTransform": "capitalize",
                },
              ]
            }
          >
            special-defense
          </Text>
          <Text
            size="small"
            style={
              Array [
                Object {
                  "color": "rgb(33, 26, 29)",
                  "fontSize": 16,
                  "fontWeight": "normal",
                },
              ]
            }
          >
            65
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
                "height": 30,
                "justifyContent": "space-between",
                "width": "48%",
              },
            ]
          }
          testID="pokemon-stat"
        >
          <Text
            size="small"
            style={
              Array [
                Object {
                  "color": "rgb(33, 26, 29)",
                  "fontSize": 16,
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "600",
                  "textTransform": "capitalize",
                },
              ]
            }
          >
            speed
          </Text>
          <Text
            size="small"
            style={
              Array [
                Object {
                  "color": "rgb(33, 26, 29)",
                  "fontSize": 16,
                  "fontWeight": "normal",
                },
              ]
            }
          >
            35
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
